glean_parser/util.py

```diff
@@ -66,13 +66,40 @@ def remove_implicit_resolver(cls, tag_to_remove):
 _NoDatesSafeLoader.remove_implicit_resolver("tag:yaml.org,2002:timestamp")
 
 
+class SafeLineLoader(_NoDatesSafeLoader):
+    """
+    Map line number to ping and metric yaml nodes.
+
+    There are a few specific places we can map the line number to:
+    only add line number to ping/metric node, and map it at the
+    *property* level so that it won’t fail schema validation.
+    Since the file structures for pings.yaml and metrics.yaml
+    are different, and yamllint is run on the input which enforces
+    2-space tab width, the logic to find the right nodes to map
+    to is defined in construct_mapping().
+    """
+
+    def construct_mapping(self, node, deep=False):
+        mapping = super(SafeLineLoader, self).construct_mapping(node, deep=deep)
+
+        for value in node.value:
+            if value[0].value != "$schema":
+                # pings.yaml
+                if node.start_mark.column == 2 and value[0].value == "description":
+                    mapping["defined_in"] = {"line": str(node.start_mark.line)}
+                # metrics.yaml
+                if node.start_mark.column == 4 and value[0].value == "type":
+                    mapping["defined_in"] = {"line": str(node.start_mark.line)}
+        return mapping
```

Comment:
Thanks for your patience with this PR -- I think it's turned out to be more complicated than anyone expected.

I think the need to recursively delete all keys called `defined_in` is a problem -- what if a metric or ping name is called `defined_in`?  It would get silently deleted.

I think the root of the issue is that this is attaching data to the dictionary, but it's really something different -- it doesn't represent anything "literally" coming from the file.  A bunch of the other weirdness flows out from this, too.

I think, instead, what we should do is use a subclass of `dict` so we can attach this line number metadata without it being part of the dictionary itself.

Using this instead seems to work:

```
class SafeLineLoader(_NoDatesSafeLoader):
    pass


def _construct_mapping_adding_line(loader, node):
    loader.flatten_mapping(node)
    mapping = DictWithMetadata(loader.construct_pairs(node))
    mapping.defined_in = {"line": node.start_mark.line}
    return mapping


SafeLineLoader.add_constructor(
    yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG, _construct_mapping_adding_line
)
```

This does mean that we'll also have to change the place where metrics are built slightly as well (in `make_metric`), so that we pull this data back out.  At the end of `make_metric`:

```
        return cls.metric_types[metric_type](
            category=category,
            name=name,
            defined_in=getattr(metric_info, "defined_in", None),
            _validated=validated,
            _config=config,
            **metric_info,
        )
```

Then, we shouldn't need to do anything special to validate the YAML loaded in (the JSON schema validator will safely ignore the extra information, because it's an *attribute* on the dictionary, not a *key*).

We'll still need to do the thing where we delete the `defined_in` key in `_serialize_input`, but we don't need to do the "delete it at any level" thing, so it should be safe.

I think this also addresses the hard coding of "description" and "type" as special keys to identify where metrics/ping appear in the file, too.

Sorry to come up with a completely different solution so late in the game -- sometimes these things don't become clear until they are worked all the way through, so no wasted effort here at all.